In [1]:
# import necessary libraries
import math
from pathlib import Path
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score, roc_curve, auc
import matplotlib.pylab as plt
import numpy as np 

from dmba import regressionSummary, classificationSummary, liftChart, gainsChart
import seaborn as sns
from sklearn.metrics import confusion_matrix,f1_score
from sklearn import preprocessing

# Q1

In [2]:
# read the dataset
df_q1 = pd.read_excel(r"D:\OneDrive - NITT\Custom_Download\regression_metric_data.xlsx")

In [3]:
# get true and predicted value
y_true = df_q1['Y_actual']
y_pred = df_q1['Y_predicted']

# call regressionSummary() function to print Regression statistics
regressionSummary(y_true,y_pred)


Regression statistics

                      Mean Error (ME) : 0.3077
       Root Mean Squared Error (RMSE) : 2.0096
            Mean Absolute Error (MAE) : 1.6538
          Mean Percentage Error (MPE) : -17.3397
Mean Absolute Percentage Error (MAPE) : 61.8269


In [4]:
df_q1.min() # check for Minimum

Y_actual       1
Y_predicted    1
dtype: int64

In [5]:
df_q1.max() # check for Maximum

Y_actual       5
Y_predicted    5
dtype: int64

- Though "Mean Error",RMSE,MAE is small because values are in the range of 1 to 5, but overall the model is giving poor performance, that can be concluded from huge MAPE and negative Mean Percentage Error (MPE)

# Q2

In [6]:
# read the dataset
df_q2 = pd.read_csv(r"D:\OneDrive - NITT\Custom_Download\Assignment_2_Q2.csv")

In [7]:
# get true and predicted value
y_true = df_q2['Actual']
y_pred = df_q2['Predicted']

# print Confusion Matrix using classificationSummary
print(classificationSummary(y_true,y_pred))

Confusion Matrix (Accuracy 0.5167)

       Prediction
Actual  0  1
     0 27 24
     1  5  4
None


In [8]:
# get true negative ,true positive ,false negative and false positive
tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()

In [9]:
tp #  true positive, the model correctly predicts the positive class

4

In [10]:
tn # true negative, the model correctly predicts the negative class

27

In [11]:
# false positive, It should be predicted as Negative but 
# the model incorrectly predicts the positive class
fp 

24

In [12]:
# false negative, It should be predicted as positive but 
# the model incorrectly predicts the negative class

fn 

5

In [13]:
# Specificity or true negative rate
TNR = tn/(tn+fp) 

# Sensitivity, recall, or true positive rate
TPR = tp/(tp+fn)

# Overall accuracy
ACC = (tp+tn)/(tp+fp+fn+tn)

# Precision 
PPV = tp/(tp+fp)

# F1 Score
F1_Score = 2*(PPV*TPR)/(PPV+TPR)

# print the result
print("Specificity:",round(TNR,3))
print("Sensitivity:",round(TPR,3))
print("accuracy:",round(ACC,3))
print("F1_Score:",round(F1_Score,3))

Specificity: 0.529
Sensitivity: 0.444
accuracy: 0.517
F1_Score: 0.216


- Based on Confusion Matrix, i can say that out of 51 time (27+24) class-0, only 27 times its getting correctly predicted. Similarly out of 9 time class-1 (5+4), only 4 times its getting correctly predicted.

- Based on the accuracy score of mere 0.517, it is safe to conclude that the model was not performing well i.e its probability of correctness is just half i.e 51.7%